src/docker.rs

```diff
@@ -0,0 +1,190 @@
+    log::*,
+    std::{
+        env,
+        error::Error,
+        fs,
+        path::{Path, PathBuf},
+        process::{Command, Output, Stdio},
+    },
+};
+
+pub struct DockerConfig {
+    pub base_image: String,
+    pub image_name: String,
+    pub tag: String,
+    pub registry: String,
+    deploy_method: DeployMethod,
+}
+
+impl DockerConfig {
+    pub fn new(
+        base_image: String,
+        image_name: String,
+        tag: String,
+        registry: String,
+        deploy_method: DeployMethod,
+    ) -> Self {
+        DockerConfig {
+            base_image,
+            image_name,
+            tag,
+            registry,
+            deploy_method,
+        }
+    }
+
+    pub fn build_image(
+        &self,
+        solana_root_path: &Path,
+        validator_type: &ValidatorType,
+    ) -> Result<(), Box<dyn Error>> {
+        match validator_type {
+            ValidatorType::Bootstrap => (),
+            ValidatorType::Standard | ValidatorType::RPC | ValidatorType::Client => {
+                return Err(format!(
+                    "Build docker image for validator type: {validator_type} not supported yet"
+                )
+                .into());
+            }
+        }
+        let image_name = format!("{validator_type}-{}", self.image_name);
+        let docker_path = solana_root_path.join(format!("docker-build/{validator_type}"));
+        match self.create_base_image(solana_root_path, image_name, &docker_path, validator_type) {
+            Ok(res) => {
+                if res.status.success() {
+                    info!("Successfully created base Image");
+                    Ok(())
+                } else {
+                    error!("Failed to build base image");
+                    Err(String::from_utf8_lossy(&res.stderr).into())
+                }
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    fn create_base_image(
+        &self,
+        solana_root_path: &Path,
+        image_name: String,
+        docker_path: &PathBuf,
+        validator_type: &ValidatorType,
+    ) -> Result<Output, Box<dyn Error>> {
+        self.create_dockerfile(validator_type, docker_path, None)?;
+
+        trace!("Tmp: {}", docker_path.as_path().display());
+        trace!("Exists: {}", docker_path.as_path().exists());
+
+        // We use std::process::Command here because Docker-rs is very slow building dockerfiles
+        // when they are in large repos. Docker-rs doesn't seem to support the `--file` flag natively.
+        // so we result to using std::process::Command
+        let dockerfile = docker_path.join("Dockerfile");
+        let context_path = solana_root_path.display().to_string();
+
+        let progress_bar = new_spinner_progress_bar();
+        progress_bar.set_message(format!("{BUILD}Building {validator_type} docker image...",));
+
+        let command = format!(
+            "docker build -t {}/{image_name}:{} -f {dockerfile:?} {context_path}",
```

Comment:
Question on using `{dockerfile:?}` here: the documentation says to use the `Debug` formatter if you want to escape the path at https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.display

Is using `display()` no good for your use-case?

Reply:
did not realize that. will update.